src/main/java/com/icthh/xm/ms/entity/service/XmEntitySpecService.java

```diff
@@ -131,27 +134,58 @@
     @SneakyThrows
     @IgnoreLogginAspect
     public void onRefresh(String updatedKey, String config) {
-        String specificationPathPattern = applicationProperties.getSpecificationPathPattern();
+
         try {
-            String tenant = matcher.extractUriTemplateVariables(specificationPathPattern, updatedKey).get(TENANT_NAME);
-            if (StringUtils.isBlank(config)) {
-                types.remove(tenant);
-                return;
-            }
-            XmEntitySpec spec = mapper.readValue(config, XmEntitySpec.class);
-            Map<String, TypeSpec> value = toTypeSpecsMap(spec);
-            types.put(tenant, value);
-            entityCustomPrivilegeService.updateCustomPermission(value, tenant);
-            log.info("Specification was for tenant {} updated", tenant);
+            String pathPattern = getPathPattern(updatedKey);
+            String tenant = matcher.extractUriTemplateVariables(pathPattern, updatedKey).get(TENANT_NAME);
+            updateByFileState(updatedKey, config, tenant);
+            Map<String, TypeSpec> tenantEntitySpec = updateByTenantState(tenant);
+            entityCustomPrivilegeService.updateCustomPermission(tenantEntitySpec, tenant);
+            log.info("Specification was for tenant {} updated from file {}", tenant, updatedKey);
         } catch (Exception e) {
             log.error("Error read xm specification from path " + updatedKey, e);
         }
     }
 
+    private String getPathPattern(String updatedKey) {
+        String specificationPattern = applicationProperties.getSpecificationPathPattern();
+        String specificationFolderPattern = applicationProperties.getSpecificationFolderPathPattern();
+        String pathPattern = null;
+        if (matcher.match(specificationPattern, updatedKey)) {
+            pathPattern = specificationPattern;
+        } else if (matcher.match(specificationFolderPattern, updatedKey)) {
+            pathPattern = specificationFolderPattern;
+        }
+        return pathPattern;
+    }
+
+    private Map<String, TypeSpec> updateByTenantState(String tenant) {
+        var tenantEntitySpec = new LinkedHashMap<String, TypeSpec>();
```

Comment:
can we use types.computeIfAbsent() here ?

Reply:
i don't understand, is I see no